kv/kvserver: TestReliableIntentCleanup failed

kv/kvserver.TestReliableIntentCleanup [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3120314&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3120314&tab=artifacts#/) on master @ [b6598d0101dc76c284e14f7809338ae14ce07f4f](https://github.com/cockroachdb/cockroach/commits/b6598d0101dc76c284e14f7809338ae14ce07f4f):


```
=== RUN   TestReliableIntentCleanup
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReliableIntentCleanup808306975
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestReliableIntentCleanup
    intent_resolver_integration_test.go:686: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReliableIntentCleanup808306975
--- FAIL: TestReliableIntentCleanup (684.84s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true
            --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true (372.38s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000
        --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000 (427.86s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true
    --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true (477.76s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push
    intent_resolver_integration_test.go:614: 
        	Error Trace:	intent_resolver_integration_test.go:404
        	            				intent_resolver_integration_test.go:614
        	            				intent_resolver_integration_test.go:674
        	            				subtest.go:34
        	Error:      	found stale intents
        	Test:       	TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push
        	Messages:   	count=1329 first={{"key\x00\rߠl"} id=a185e422 key="key\x00\x06\x84\x0f\xc4" pri=0.00548926 epo=0 ts=1624645357.969730944,0 min=1624645357.969730944,0 seq=213} last={{"key\x00\x0f\xff\xbe\x9f"} id=a185e422 key="key\x00\x06\x84\x0f\xc4" pri=0.00548926 epo=0 ts=1624645357.969730944,0 min=1624645357.969730944,0 seq=8871}
                        --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push (100.78s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel
                    --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel (202.64s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true
                --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true (369.00s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestReliableIntentCleanup PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReliableIntentCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>



Jira issue: CRDB-8279

Epic CRDB-27234

--- a/pkg/kv/kvserver/intent_resolver_integration_test.go
+++ b/pkg/kv/kvserver/intent_resolver_integration_test.go
@@ -189,10 +189,11 @@ func TestContendedIntentWithDependencyCycle(t *testing.T) {
 func TestReliableIntentCleanup(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.WithIssue(t, 66895, "Flaky due to non-deterministic txn cleanup")
+
 	skip.UnderShort(t) // takes 294s
 	skip.UnderRace(t, "timing-sensitive test")
-	skip.UnderStress(t, "memory-hungry test")
+	skip.UnderDeadlock(t, "timing-sensitive test")
+	//skip.UnderStress(t, "memory-hungry test")
 
 	prefix := roachpb.Key([]byte("key\x00"))
 
@@ -447,18 +448,18 @@ func TestReliableIntentCleanup(t *testing.T) {
 				// meanwhile, returning when heartbeat was aborted.
 				abortedC := abortHeartbeat(t, txnKey)
 				readyC := blockPut(t, txnKey)
-				require.NoError(t, tc.Stopper().RunAsyncTask(ctx, "unblock", func(_ context.Context) {
+				go func() {
 					<-abortedC
 					unblockC := <-readyC
 					time.Sleep(100 * time.Millisecond)
 					close(unblockC)
-				}))
+				}()
 				close(abortErrC) // can't error
 
 			case "push":
 				// Push txn with a high-priority write once put is blocked.
 				readyC := blockPut(t, txnKey)
-				require.NoError(t, tc.Stopper().RunAsyncTask(ctx, "push", func(ctx context.Context) {
+				go func() {
 					unblockC := <-readyC
 					defer close(unblockC)
 					defer close(abortErrC)
@@ -482,8 +483,8 @@ func TestReliableIntentCleanup(t *testing.T) {
 						abortErrC <- err
 						return
 					}
-					time.Sleep(100 * time.Millisecond)
-				}))
+					time.Sleep(100 * time.Millisecond) // sleep before defers
+				}()
 
 			case "":
 				close(abortErrC)
@@ -496,12 +497,12 @@ func TestReliableIntentCleanup(t *testing.T) {
 			// evaluated in Raft.
 			if spec.finalize == "cancelAsync" {
 				readyC := blockPutEval(t, txnKey)
-				require.NoError(t, tc.Stopper().RunAsyncTask(ctx, "cancel", func(_ context.Context) {
+				go func() {
 					unblockC := <-readyC
 					defer close(unblockC)
 					cancel()
-					time.Sleep(100 * time.Millisecond)
-				}))
+					time.Sleep(100 * time.Millisecond) // sleep before defers
+				}()
 			}
 
 			// Write numKeys KV pairs in batches of batchSize as a single txn.
@@ -569,15 +570,15 @@ func TestReliableIntentCleanup(t *testing.T) {
 					break
 				} else if spec.abort != "" {
 					require.Error(t, err)
-					if errors.As(err, retryErr) {
+					if errors.As(err, &retryErr) {
 						require.True(t, retryErr.PrevTxnAborted())
 					} else {
 						require.Fail(t, "expected TransactionRetryWithProtoRefreshError, got %v", err)
 					}
 					break
-				} else if !errors.As(err, retryErr) {
+				} else if !errors.As(err, &retryErr) {
 					require.NoError(t, err)
-				} else if attempt >= 3 {
+				} else if attempt >= 7 {
 					require.Fail(t, "too many txn retries", "attempt %v errored: %v", attempt, err)
 				} else {
 					t.Logf("retrying unexpected txn error: %v", err)
@@ -632,6 +633,11 @@ func TestReliableIntentCleanup(t *testing.T) {
 					}
 					finalize := []interface{}{"commit", "rollback", "cancel", "cancelAsync"}
 					testutils.RunValues(t, "finalize", finalize, func(t *testing.T, finalize interface{}) {
+						// TODO(erikgrinaker): cleanup does not always work reliably with
+						// context cancellation, so we skip these for now to deflake the test.
+						if finalize == "cancel" || finalize == "cancelAsync" {
+							skip.WithIssue(t, 105615)
+						}
 						if finalize == "cancelAsync" {
 							// cancelAsync can't run together with abort.
 							testTxn(t, testTxnSpec{

